lib/block-supports/visibility.php

```diff
@@ -0,0 +1,46 @@
+<?php
+/**
+ * Block visibility block support flag.
+ *
+ * @package gutenberg
+ */
+
+/**
+ * Render nothing if the block is hidden.
+ *
+ * @param string $block_content Rendered block content.
+ * @param array  $block         Block object.
+ * @return string Filtered block content.
+ */
+function gutenberg_render_visibility_support( $block_content, $block ) {
+	$block_type = WP_Block_Type_Registry::get_instance()->get_registered( $block['blockName'] );
+
+	if (
+		! $block_type ||
+		! block_has_support( $block_type, 'visibility', true ) &&
+		// This condition is not needed in core.
+		! block_has_support( $block_type, 'blockVisibility', true )
+	) {
+		return $block_content;
+	}
+
+	// The block support key used to be `blockVisibility` and was later renamed
+	// to `visibility`. When the Gutenberg plugin is updated, hidden content could
+	// unexpectedly render on the front end if the new key is only checked.
+	// To prevent this, we continue checking the old support key too.
+	// This conditional statement is not needed in core.
+	if ( isset( $block['attrs']['metadata']['blockVisibility'] ) && false === $block['attrs']['metadata']['blockVisibility'] ) {
+		return '';
+	}
```

Comment:
Unfortunately, while this provides backwards compatibility, it also means that it is impossible to unhide older blocks that used the previous `blockVisibility` attribute via the UI.

Here's a video where I've got two paragraph blocks hidden, one using the new attribute, and one set before applying this PR. It's possible to show/hide the one using `visibility`, but the one using `blockVisibility` remains hidden irrespective of show/hide value:

https://github.com/user-attachments/assets/3b947dc7-3b2e-43d0-9435-abed680a7b5b